design-documents/graph-ql/coverage/quote.graphqls

```diff
@@ -73,7 +73,10 @@ input UpdateCartItemsInput {
 }
 
 input CartItemUpdateInput {
-    cart_item_id: Int!
+    # Implementation Note: The resolver should accept either `cart_item_id` or `cart_item_uid`,
+    # but _must_ return an error when both fields are provided in a request
+    cart_item_id: Int! @deprecated(reason: "Use `CartItemUpdateInput.cart_item_uid` instead")
+    cart_item_uid: ID! @doc(description: "Unique Identifier from objects implementing `CartItemInterface`")
```

Comment:
Maybe I'm wrong but I think there is a problem here. You cannot have both arguments mandatory, otherwise the schema will reject any request if both `cart_item_id` and `cart_item_uid` are not set. I quickly did a test with GraphiQL in Chrome with a Magento schema, and I'm for example forced to set `cart_id` in `AddSimpleProductsToCartInput` (it's also declared with `!`). This is going to be problematic, especially if you want to return an error when both fields are provided.

You could resolve this by adding an extra input type like
```
input CartItemIdInput {
    cart_item_id: Int @deprecated(reason: "Use `CartItemUpdateInput.cart_item_uid` instead")
    cart_item_uid: ID @doc(description: "Unique Identifier from objects implementing `CartItemInterface`")
}
```
and make this one mandatory like in
```
input CartItemUpdateInput {
    cart_item_id: Int
    cart_item_id_input: CartItemIdInput!
}
```
but this would make everything complicated and awkward. The easy fix is to remove the mandatory parts for both fields, but this is also not very nice, and you cannot only enforce the new ID or you'll break backwards-compatibility.

Reply:
I guess both as mandatory would also be a problem for backwards compatibility. Older clients will not be sending cart_item_uid, right?

Reply:
> Maybe I'm wrong but I think there is a problem here

You are not wrong - good catch! They're both marked as non-nullable so this fails the query at the validation step, before it ever hits the resolvers.

This is unfortunately one of those problems that's hard to model without polymorphic input types, [which there has been a _lot_ of discussion/debate about for awhile](https://github.com/graphql/graphql-spec/blob/d0b76b0bd285b61df2210a57144b9a70d8d7dc4e/rfcs/InputUnion.md)

How would you feel about the following change, which is just re-iterating your idea:

1. Both fields (`cart_item_id` and `cart_item_uid`) are made nullable (removing `!`) in schema
2. Validation of required fields (i.e. the mandate that one exists) is validated in the resolver for the mutation

This would provide the following benefits:
1. Backwards-compatible for existing clients
1. Impossible to pass both IDs without an error
1. Impossible to pass no IDs without an error
1. (Mostly) Safe to make `cart_item_uid` non-nullable (`!`) later when we remove `cart_item_id`

But with the following downsides:
1. We lose some useful information in the schema, because we're side-stepping the type system and using our own validation in a way that isn't easy to introspect.

Reply:
+1 for this solution, I don't think there is any better solution that doesn't introduce its own set of issues or backwards-incompatible problems. And thanks for the link on `inputunion`, I didn't know about that, indeed very interesting reading.

Reply:
Fixed in a66f16687c05